### ISSUE TITLE
DIALOG and TM modules correlation when there is a race between CANCELs and 200 OKs

### ISSUE THREAD

**Issue by vladpaiu** [OpenSIPS/opensips#1388]
Hello,

While in an ideal world the UACs should take care of races between 200OK and CANCELs, real life is nothing like that, and this often leads to hanged calls and disputes - since nobody is really sending a BYE for a 200 OK after they've sent a CANCEL.

It would be nice to have the TM and DIALOG module correlate on this issue, and be able to auto-kill such a call, that is established due to a race condition between a 200Ok and a CANCEL. For example, if the call is CANCELed but a 200OK is received, the dialog module could internally auto-kill the dialog ( maybe by setting the DLG timeout to a low value ).
Also, if the CANCEL comes in for a 200OKed transaction, then again, the TM module could signal this to the dialog module, which should internally auto-kill the dialog.
This could be controlled by a dialog modparam, allowing to auto-kill such race condition dialogs ( default modparam value would be to not interfere at all - keep the regular proxy behavior ).

Let me know if this sounds like a reasonable solution to the problem - if yes, I can return with a pull request.

Best Regards,
Vlad

**Comment by rrb3942** [comment]
I think you can do something similar to this currently by checking if the transaction was cancelled when you receive the 200 OK and then set the dialog timeout to 0, causing it to hangup the call. This is something I've tried in the past, but from my experience it causes more problems than it solves.

I'm also pretty sure this would break RFC compliance. A clip from Section 9.2 of RFC 3261 (Server behaviour for a CANCEL):

   If the transaction
   for the original request still exists, the behavior of the UAS on
   receiving a CANCEL request depends on whether it has already sent a
   final response for the original request.  If it has, the CANCEL
   request has no effect on the processing of the original request, no
   effect on any session state, and no effect on the responses generated
   for the original request.

So for a race condition the UAC should be deciding if it wants to continue the call or end it. You might want this for a notably buggy endpoint, but I think if it was a global on/off it might be problematic.

**Comment by vladpaiu** [comment]
Hello,

I've also fixed the described issue by checking at 200OK time ( in the script ) if there was any CANCEL for the respective call, and if a CANCEL had been received, I'd set the $DLG_timeout to 0 - hackish, but gets the job done ; don't think anybody can implement this mechanism in the script in order to mitigate such races.

RFC compliance is not something I'm particularly interested in - RFC UA compliance is sometimes way worse and the only way to fix it is to have a proxy that slightly bends the rules in order to accommodate buggy clients ( eg - see fix_route_dialog() ).

In short, the UAC should decide, but the UAC rarely decides - if ever it is coded smart enough to even know what happened.

Regards,
Vlad

**Comment by bogdan-iancu** [comment]
Hey @vladpaiu , could you please update here as per our yesterday discussion?

**Comment by vladpaiu** [comment]
Hello,

Just committed a5f4955c3de9874fde7c59746d14cd8e4602df64 , which should address the race both ways ( 2OO OK first or CANCEL first ) - it adds a new 'E' flag to create_dialog() that controls whether the dialog is auto-terminated when the race occurs.

Regards,
Vlad

**Comment by wphowell** [comment]
We are experience this same crash, but when reviewing the create_dialog documentation, there is no "E" flag mentioned.  If this was added to resolve this issue, and it's a condition that can crash OpenSIPS, why isn't it documented?  We don't want to add an undocumented flag.
